### PR TITLE
Add a Products by Rating filter block

### PR DESCRIPTION
**Pull request by danieldudzic**
This is an attempt at adding a **Products by Rating** filter block.

For now, I have focused on the core functionality, and there are several pending items on the to-do list:

#### To-do

- [ ] Add support for **Display Options**: Product count, Show 'Apply filters' button, Color
- [ ] Fix the editor preview display
- [ ] Migrate the block from directly using the `Rating` component to utilizing the `CheckboxList` component (this will also make the UX match the preferred option ( pdnLyh-1sR-p2#comment-1044 ))
- [ ] Investigate why the Active Filters block `Clear All` button doesn't work correctly (doesn't clear Rating filters on a Block Shop page)
- [ ] Provide the legacy Filter Products by Rating widget deprecation
- [ ] Add E2E tests
- [ ] Add the ability for the `Rating` component to support other contexts (for example the All Products block)

### Screenshots
New_shop_–_ratings

### Testing
- Activate a block theme & setup WooCommerce with products that have some ratings
- Create a classic shop page and a separate page using the `All Products` block
- Add the `Filter Products by Rating` block and `Active Filters` block on both pages
- Check that `Filter Products by Rating` block is correctly:
  - Displaying the star rating buttons within the filter block
  - Displaying the active state (stars using the Primary color and the product count are in bold)
  - Filtering the products in the product loop (on both Classic and block shop pages)
  - Clearing after getting removed via `Active Filters` remove button
  - Updating the URL with the rating parameter
  - Loading the rating filters via URL parameters on page reload

#### User Facing Testing

Same as the above steps

### WooCommerce Visibility

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

### Changelog

> Add Filter Products by Rating block

### CODE REVIEW

File: assets/js/blocks/rating-filter/block.json
Comment (Aljullu) [inline assets/js/blocks/rating-filter/block.json]: We are removing the `Works in combination...` part of the filter blocks description (see #7045).

File: assets/js/blocks/rating-filter/block.json
Comment (Aljullu) [inline assets/js/blocks/rating-filter/block.json]: I guess the block name should be updated to `Filter by Rating`.

File: src/BlockTypesController.php
Comment (Aljullu) [inline src/BlockTypesController.php]: Nit: I would list this block under `StockFilter` (line 185), so it's listed next to all other filter blocks.

File: assets/js/blocks/rating-filter/index.tsx
Comment (Aljullu) [inline assets/js/blocks/rating-filter/index.tsx]: Given that this is defined in `block.json`, I think you can remove it from here. We left it in the other filter blocks because when moving from one file to the other it might break existing translations for some time, but it shouldn't happen in your case because the block is new.

**Review by Aljullu — COMMENTED** [review]
This is working great, congrats for working on this big project, @danieldudzic! I did an initial review and left some minor issues below.

I also noticed the stars are not visible in the Site Editor:

imatge

I'm reporting it here but IMO this can be fixed in a follow-up PR. (Sorry if you had already identified it, maybe that's what you were referring to with _Fix the editor preview display_)

For this PR to be ready to merge, I think we need to:

- Add an experimental build gate (similar to [this one](https://github.com/woocommerce/woocommerce-blocks/blob/71001ee44511c34761e292e1571ad8b4feb16fcc/assets/js/blocks/product-query/variations/product-query.tsx#L15) and [this one](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/bin/webpack-entries.js#L61)).
- I think it would be good to fix all the inline comments I left in this PR since they are minor things and affect translations.
- I also think it would be great to work on this item: _Migrate the block from directly using the Rating component to utilizing the CheckboxList component_. If I'm understanding this properly, this will simplify the code significantly.

IMO, all other To-do items you identified can then be worked on in follow-up issues/PRs. How does it sound?

File: assets/js/blocks/rating-filter/block.json
Comment (danieldudzic) [reply assets/js/blocks/rating-filter/block.json]: > Sorry if you had already identified it, maybe that's what you were referring to with Fix the editor preview display

That's exactly it :)

> Add an experimental build gate

I'll add it today

> Migrate the block from directly using the Rating component to utilizing the CheckboxList component. If I'm understanding this properly, this will simplify the code significantly.

Yes, I think it would make a lot of sense. I'll work on that and see if this is something I can get done by the end of the week.

**Comment by danieldudzic** [comment]
@Aljullu Re:

> I also noticed the stars are not visible in the Site Editor

I have just noticed the same issue occurs for the All Products block. Would you mind having a look?

Editor__beta__‹_ratings_—_WordPress

File: assets/js/base/components/product-rating/index.tsx
Comment (Aljullu) [inline assets/js/base/components/product-rating/index.tsx]: You can remove `role="button"` and `tabIndex={ key }`. We no longer want this `<div>` to be interactive because it's now wrapper in a `<label>`, which is already clickable.

File: assets/js/base/components/product-rating/index.tsx
Comment (Aljullu) [inline assets/js/base/components/product-rating/index.tsx]: Hmmm, I don't think we need to account for plurals, here. You could replace this [`_n()`](https://developer.wordpress.org/reference/functions/_n/) with [`__()`](https://developer.wordpress.org/reference/functions/__/) (similar to line 29).

File: assets/js/base/components/product-rating/style.scss
Comment (Aljullu) [inline assets/js/base/components/product-rating/style.scss]: Just curious, is this required by a specific theme? I tried with Twenty Twenty Two and Storefront and it doesn't make any difference if I remove this rule. Same with `margin-left: unset` below. :thinking:

File: assets/js/blocks/rating-filter/style.scss
Comment (Aljullu) [inline assets/js/blocks/rating-filter/style.scss]: Is there any reason to set `cursor: pointer` only to active ratings? I think we should set it in all of them, no matter if they are active or not.

File: assets/js/blocks/rating-filter/style.scss
Comment (Aljullu) [inline assets/js/blocks/rating-filter/style.scss]: Heads-up that this color might not be available in all themes (ie: Storefront): I think we should provide a fallback one.

**Review by Aljullu — COMMENTED** [review]
Thanks for applying the changes and updating this PR, @danieldudzic!

> I have just noticed the same issue occurs for the All Products block. Would you mind having a look?

You are right! Could you fill a separate issue for that? I don't think we should block this PR if this is affecting other blocks.

I left some inline comments below. :slightly_smiling_face:

**Comment by danieldudzic** [comment]
> Thanks for applying the changes and updating this PR, @danieldudzic!
> 
> > I have just noticed the same issue occurs for the All Products block. Would you mind having a look?
> 
> You are right! Could you fill a separate issue for that? I don't think we should block this PR if this is affecting other blocks.

Issue created: https://github.com/woocommerce/woocommerce-blocks/issues/7167

File: assets/js/blocks/rating-filter/style.scss
Comment (Aljullu) [inline assets/js/blocks/rating-filter/style.scss:23]: Given that the `input` is inside `label`, you can remove the first selector: applying [`cursor: pointer`](https://developer.mozilla.org/en-US/docs/Web/CSS/cursor#formal_definition) to the label will also make the input have the same cursor.
```suggestion
	.wc-block-rating-filter .wc-block-rating-filter-list li label {
```

**Review by Aljullu — APPROVED** [review]
This is looking great and it can be already merged, so pre-approving. This is exciting! :clap: 

Some things to do before merging:
* In the PR description, `WooCommerce Visibility` should be set to `Experimental` given that this is gated with `isExperimentalBuild()`. Then, it probably makes sense to add the `skip-changelog` label to make sure this is not mentioned in the changelog.
* You will need to rebase your branch before being able to merge.
* I left a small comment about a CSS selector that you can remove.

And a couple of follow-ups:

* In recent PRs, we added [loading skeletons](https://github.com/woocommerce/woocommerce-blocks/pull/6990) to filter blocks. We will need to do the same for the Products by Rating block. @danieldudzic can you create an issue for this?
* I noticed that some elements get a numeric id (ie: `id="5"`). I think that's problematic because ids need to be unique in the page and it's easy to introduce a clash at some point if we only use a number as the id. Ideally we should make sure those ids are really unique, either in the form of `wc-blocks-products-by-rating-5` (or something along these lines) or [using `uuid()` as in GB](https://github.com/WordPress/gutenberg/blob/c4787412eef5959d8f2e519375352b8f4fed231a/test/native/integration-test-helpers/initialize-editor.js#L28). I know this is something that is inherited from the Filter by Attribute so it probably makes sense working on both blocks at the same time. However, could you create an issue as well so we don't forget?

File: assets/js/blocks/rating-filter/style.scss
Comment (danieldudzic) [reply assets/js/blocks/rating-filter/style.scss:23]: In theory, yes, but in practice removing the input selector breaks the `cursor: pointer` for the inputs 😅

**Comment by danieldudzic** [comment]
> I noticed that some elements get a numeric id (ie: `id="5"`). I think that's problematic because ids need to be unique in the page and it's easy to introduce a clash at some point if we only use a number as the id. Ideally we should make sure those ids are really unique, either in the form of `wc-blocks-products-by-rating-5` (or something along these lines) or [using `uuid()` as in GB](https://github.com/WordPress/gutenberg/blob/c4787412eef5959d8f2e519375352b8f4fed231a/test/native/integration-test-helpers/initialize-editor.js#L28). I know this is something that is inherited from the Filter by Attribute so it probably makes sense working on both blocks at the same time. However, could you create an issue as well so we don't forget?

I have now created an issue for this, here: https://github.com/woocommerce/woocommerce/issues/42579

**Comment by danieldudzic** [comment]
> In recent PRs, we added [loading skeletons](https://github.com/woocommerce/woocommerce-blocks/pull/6990) to filter blocks. We will need to do the same for the Products by Rating block. @danieldudzic can you create an issue for this?

I have now created an issue for this, here: https://github.com/woocommerce/woocommerce-blocks/issues/7208

File: assets/js/blocks/rating-filter/block.tsx
Comment (Aljullu) [inline assets/js/blocks/rating-filter/block.tsx]: I'm not a TS expert so there might be a better solution. But after playing with this a bit, I think you could set the type here and then remove the `orderedRatings.filter()` below.

```suggestion
	let displayedOptions: DisplayOption[] = [];
```

You will also need to define `DisplayOption` in `rating-filter/types.ts` (and import it from this file). That `DisplayOption` type could be the same as the one in `attribute-filter/types.ts` but only with `label` and `value` properties.

**Review by Aljullu — APPROVED** [review]
Thanks for continuing improving this PR, @danieldudzic! I left a comment about the TS definition of `displayedOptions`.

Besides that, I noticed that the placeholder title is strangely big compared to the other filter blocks:



Can you reproduce as well? I don't think this is blocking this PR to be merged, so IMO you can fix it on a follow-up PR so we don't block this one from getting into `trunk`.

**Comment by danieldudzic** [comment]
> Besides that, I noticed that the placeholder title is strangely big compared to the other filter blocks

Thank you for the review!

I'm not seeing the issue with the large placeholder:
New_shop_–_ratings

New_shop_–_ratings

File: assets/js/blocks/rating-filter/block.tsx
Comment (gigitux) [reply assets/js/blocks/rating-filter/block.tsx]: Regarding this topic, we could avoid TS errors, if we avoid mutability:

```
	const orderedRatings =
		! filteredCountsLoading &&
		objectHasProp( filteredCounts, 'rating_counts' ) &&
		Array.isArray( filteredCounts.rating_counts )
			? [ ...filteredCounts.rating_counts ].reverse()
			: [];

	const displayedOptions = orderedRatings
		.filter(
			( item ) => isObject( item ) && Object.keys( item ).length > 0
		)
		.map( ( item ) => {
			return {
				label: (
					<Rating
						className={
							productRatingsArray.includes(
								item?.rating?.toString()
							)
								? 'is-active'
								: ''
						}
						key={ item?.rating }
						rating={ item?.rating }
						ratedProductsCount={ item?.count }
					/>
				),
				value: item?.rating?.toString(),
			};
		} );
```

I suggest to create a dedicated function for doing the check regarding `orderedRatings` and giving an explicit name to the function

**Review by gigitux — CHANGES_REQUESTED** [review]
I'm able to reproduce the big placeholder, but it makes sense to create a follow-up issue 👍 

I think that before merging this PR, it is necessary fixing the TS type error. (I suggested to you how we could fix the issue, let me know what you think!)

**Review by gigitux — APPROVED** [review]
LGTM!

Thanks for having applied the fix!

**Pull request merged**
